Combine upgrade testing and many improvements

desrosj opened the pull request:
There's a lot of reshuffling here. In summary:
- I made the database service more generic. This will make it easier to support different database types later on.
- I combined all of the upgrade testing into one workflow.
- Made general GHA improvements and consistency changes.

desrosj committed:
Include a concurrency configuration to limit duplicate workflows.

desrosj committed:
Merge all upgrade workflows and include improvements within install tests.

desrosj committed:
Only run `push` testing on primary branch.

desrosj commented on .github/workflows/install-testing.yml:
```diff
@@ -0,0 +1,86 @@
+name: Installation Tests
+
+on:
+  push:
```
Anytime a pull request or push to the primary branch changes the workflow file, we should run the tests to confirm that everything works as expected after the changes.

desrosj committed:
Avoid hardcoding 6.3 as a version, and allow latest or nightly.

desrosj commented on .github/workflows/install-testing.yml:
```diff
@@ -0,0 +1,86 @@
+      new-version:
+        description: 'The version to test installing.'
+        type: string
+        default: 'latest'
```
I wanted to avoid making the version required, but also didn't want to have the latest version hard coded here. This allows the default to be whatever the most recent release is (WP CLI default behavior), any `x.y` or `x.y.z` version, or even `nightly`.

desrosj commented on .github/workflows/install-testing.yml:
```diff
@@ -0,0 +1,86 @@
+        default: 'latest'
+
+# Cancels all previous workflow runs for pull requests that have not completed.
+concurrency:
```
This prevents multiple workflows from running simultaneously on the same PR branch, or for the exact same commit SHA.

desrosj commented on .github/workflows/upgrade-testing.yml:
```diff
@@ -0,0 +1,167 @@
+  #
+  # Spawns upgrade testing for WordPress 6.x versions with MySQL.
+  #
+  upgrade-tests-wp-6x-mysql:
```
I chose to make each job DB type specific because if you include both in 1 strategy matrix:
- it will definitely exceed the maximum of 256 jobs
- The list of excludes could end up being long.
Having it separate just makes it more clear when reading as well.

desrosj committed:
Correct variable name.

desrosj committed:
Use the correct repository and branch.

desrosj commented on .github/workflows/upgrade-testing-run.yml:
```diff
@@ -0,0 +1,94 @@
+    timeout-minutes: 20
+
+    services:
+      database:
```
I changed the service name to `database` to make it easier to add other database types later. We should be able to add MariaDB testing fairly easily.

desrosj commented on .github/workflows/upgrade-testing-run.yml:
```diff
@@ -0,0 +1,94 @@
+        run: |
+          sudo systemctl start ${{ inputs.db-type }}
+
+      - name: Download WordPress ${{ inputs.wp }}
```
I separated all of these commands into their own steps. I try to make each step as short as possible. It makes it easier to see where failures occur within the workflow.

desrosj commented on .github/workflows/upgrade-testing.yml:
```diff
@@ -0,0 +1,167 @@
+  #
+  upgrade-tests-wp-6x-mysql:
+    name: ${{ matrix.wp }} to ${{ inputs.new-version && inputs.new-version || 'latest' }}
+    uses: WordPress/wordpress-upgrade-tester/.github/workflows/upgrade-testing-run.yml@combine-upgrade-test-workflows
```
These will need to be changed to use `@trunk` for `wordpress-develop` once merged into Core.

desrosj commented on .github/workflows/upgrade-testing.yml:
```diff
@@ -0,0 +1,167 @@
+  #
+  # Spawns upgrade testing for WordPress 4.x versions on PHP 7.x with MySQL.
+  #
+  upgrade-tests-wp-4x-php-7x-mysql:
```
Separating out the PHP 7.x and 8.x strategies:
- eliminates the need for any `exclude` or `include` statements.
- Ensures each group falls under the 256 maximum job limit for a strategy matrix.

desrosj commented on .github/workflows/upgrade-testing.yml:
```diff
@@ -0,0 +1,167 @@
+    strategy:
+      fail-fast: false
+      matrix:
+        os: [ 'ubuntu-latest' ]
```
I guess we don't need to have `os` and `db-type` in the matrix and hard coded as inputs. Realistically, though, these will not be expanded because of the limitations. But I think it may be better to leave in case something else is added later where these would need to be referenced (like in the name perhaps).

costdev replied:
Yeah it's unfortunate that we're tied to Linux because of service containers. It's possible to run tests on other OS if we did individual install/setup of the database, etc. Even if we did that, it wouldn't necessarily be done in this file.

desrosj replied:
It's possible that having fewer matrix variables could make the workflow UI and status reporting faster. May be worth trying by eliminating the `os` and `db-type` variables and hardcoding them to test.

desrosj commented on .github/workflows/upgrade-testing.yml:
```diff
@@ -0,0 +1,167 @@
+        php: [ '7.0', '7.1', '7.2', '7.3', '7.4', '8.0', '8.1', '8.2' ]
+        db-type: [ 'mysql' ]
+        db-version: [ '5.7', '8.0' ]
+        wp: [ '6.0', '6.1', '6.2' ]
```
Since 6.x supports 7.0-8.2, we may need to split this one into a new job at 6.5 to keep the number low enough.

costdev replied:
Yeah we'll likely need to keep a close eye on the PHP minimum version support. Although this may change again in the coming releases as PHP 7.0 and 7.1 support continues to drop. We may find ourselves in a "sweet spot" right now, but your point stands as a consideration for the near/far future

desrosj committed:
Correct input name.

desrosj committed:
Improve upgrade testing to allow `latest` and `nightly`.

desrosj commented on .github/workflows/upgrade-testing.yml:
```diff
@@ -0,0 +1,168 @@
+    with:
+      os: ${{ matrix.os }}
+      php: ${{ matrix.php }}
+      db-type: ${{ matrix.db-type }}
```
This naming also is consistent with what I proposed in https://github.com/WordPress/wordpress-develop/pull/5016.

costdev commented on .github/workflows/upgrade-testing-run.yml:
```diff
@@ -0,0 +1,95 @@
+        run: |
+          sudo systemctl start ${{ inputs.db-type }}
+
+      - name: Download WordPress ${{ inputs.wp }}
```
For merging to Core, we _may_ need to consider where this downloads to

desrosj replied:
Could you explain?

costdev replied:
If this is merged into `wordpress-develop`, it'll download WordPress to the root directory of the repo. This largely won't be an issue, but we may end up having a clash between repo files and test files. For example, we could create a directory and pass `--path=./install-test` to the WP-CLI commands.

desrosj replied:
Ah ha, gotcha. Looking at this, I don't think we actually need to checkout the repository at all. We don't rely on any files in `wordpress-develop`. WP CLI is downloading what we actually need. We can probably just remove that step from both workflows.

desrosj commented:
I tried to add some inline comments explaining big changes, but there was a lot of reshuffling so I may have missed something.

An observation on the workflow. It appears to run very slow (takes 20-30 minutes to show that the workflow has succeeded or failed). But, that's actually not the case. The jobs themselves are running very fast. The workflow _interface_ seems to have a hard time keeping up. I confirmed this by watching the organization level Actions page that shows the total number of running jobs. It spikes to our maximum allowed (500) when the workflow is created, but only stays that high for a minute or two. After that it falls back to normal, but the UI still lags behind.

While I really do prefer one workflow, if it's that much slower, we could explore one workflow per major version range (4.x, 5.x, etc.). Just no need to have a multisite and single site workflow.

Something that occurred to me is if we want to use these tests when releasing security updates, there's a few additional considerations we need to add:
- While we currently only support PHP 7.0+, all previous branches supported 5.6, and earlier than that some other 5.x versions of PHP. If we're testing say a `6.0.6`, do we need to run those tests on 5.6 as well?
- We need to add checks to ensure we're not trying to upgrade to a lower version number. If trying to test updating to `6.0.6`, we shouldn't span 6.1, 6.2, and 6.3 jobs, because I believe those would always fail to update since WP CLI would return an error.

desrosj committed:
Use correct input name.

desrosj commented:
> * While we currently only support PHP 7.0+, all previous branches supported 5.6, and earlier than that some other 5.x versions of PHP. If we're testing say a `6.0.6`, do we need to run those tests on 5.6 as well?

Additionally, do we care about also testing MySQL 5.6 on older versions that did support that version?

desrosj committed:
Pass `latest` to prevent overriding the default.

costdev commented:
If there's a delay in seeing the succeeded jobs, I agree we might consider another workflow strategy.

We need to determine the best way to handle differing:
- PHP version support
- SQL version support
- Single site and Multisite (MS has additional database tables on install, for example, which may end up broken after Core changes)

And how to represent these in as few workflow files as possible.

What if we added all the options to the matrix, and added something like this?

`if: ${{ matrix.wp < '6.3.0' && matrix.php < '7.0.0' || <something> }}`

🤔 Wouldn't handle updates to earlier versions hmm, but maybe something like:

```
inputs.new-version == 'latest' || inputs.new-version == 'nightly' || inputs.new-version > matrix.wp
```

desrosj committed:
Add PHP 8.3 testing.

desrosj committed:
Documentation updates.

desrosj committed:
Don't checkout the repository.

This shouldn't be necessary since we're downloading through WP-CLI and testing the built packages.

desrosj committed:
Update `setup-php` action.

desrosj commented:
I made a few more changes, but I think this is ready to land. Let's get it into trunk and iterate. Remaining items:
- The workflow is slow reporting results, but the workflow itself is effective and does work.
- Explore testing point releases for older major versions. Testing should stop at that version and database and PHP versions that were not supported for that release should not be tested.
- The `uses` line in the workflow needs to be changed to use the one that's included in `trunk` when committed to Core.
- We should add Slack notifications and the auto-retry when porting to Core.
- We need to explore also adding MariaDB testing.

desrosj committed:
Reference the workflow in `trunk`.
